:bomb: `Bug`: `gitlab-ce-docker` Installation Failed with Variables Type Error

### What Happened?

- config.yaml

```yaml
config:
  state:
    backend: local
    options:
      stateFile: devstream.state

vars:
  gitlabHostname: gitlab.example.com
  jenkinsHostname: jenkins.example.com
  harborHostname: harbor.example.com
  harborURL: http://harbor.example.com
  jenkinsAdminUser: admin
  jenkinsAdminPassword: changeme
  gitlabSSHPort: 30022
  gitlabHttpPort: 30080
  gitlabHttpsPort: 30443

tools:
- name: gitlab-ce-docker
  instanceID: default
  dependsOn: []
  options:
    hostname: [[ gitlabHostname ]]
    gitlabHome: /srv/gitlab
    sshPort: [[ gitlabSSHPort ]]
    httpPort: [[ gitlabHttpPort ]]
    httpsPort: [[ gitlabHttpsPort ]]
    rmDataAfterDelete: false
    imageTag: "rc"
- name: helm-installer
  instanceID: jenkins-001
  dependsOn: []
  options:
    valuesYaml: |
...
      JCasC:
        defaultConfig: true
- name: helm-installer
  instanceID: harbor-001
  dependsOn: []
  options:
    valuesYaml: |
      externalURL: [[ harborURL ]]
...
```

- logs

`dtm apply -f config.yaml`

```sh
2022-11-25 13:16:27 ℹ [INFO]  Apply started.
2022-11-25 13:16:27 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-11-25 13:16:27 ℹ [INFO]  Tool (gitlab-ce-docker/default) found in config but doesn't exist in the state, will be created.
2022-11-25 13:16:27 ℹ [INFO]  Tool (helm-installer/jenkins-001) found in config but doesn't exist in the state, will be created.
2022-11-25 13:16:27 ℹ [INFO]  Tool (helm-installer/harbor-001) found in config but doesn't exist in the state, will be created.
Continue? [y/n]
Enter a value (Default is n): y

2022-11-25 13:16:29 ℹ [INFO]  Start executing the plan.
2022-11-25 13:16:29 ℹ [INFO]  Changes count: 3.
2022-11-25 13:16:29 ℹ [INFO]  -------------------- [  Processing progress: 1/3.  ] --------------------
2022-11-25 13:16:29 ℹ [INFO]  Processing: (gitlab-ce-docker/default) -> Create ...
2022-11-25 13:16:29 ℹ [INFO]  -------------------- [  Processing progress: 2/3.  ] --------------------
2022-11-25 13:16:29 ℹ [INFO]  Processing: (helm-installer/jenkins-001) -> Create ...
2022-11-25 13:16:29 ℹ [INFO]  Filling default config with instance: jenkins-001.
2022-11-25 13:16:29 ℹ [INFO]  Creating or updating helm chart ...
2022/11/25 13:16:30 creating 13 resource(s)
2022/11/25 13:16:30 beginning wait for 13 resources with timeout of 10m0s
2022/11/25 13:16:31 StatefulSet is not ready: jenkins/jenkins. 0 out of 1 expected pods are ready
...
2022/11/25 13:19:02 Deployment is not ready: harbor/harbor-jobservice. 0 out of 1 expected pods are ready
2022/11/25 13:19:04 release installed successfully: harbor/harbor-1.10.2
2022-11-25 13:19:04 ✔ [SUCCESS]  Tool (helm-installer/harbor-001) Create done.
2022-11-25 13:19:04 ℹ [INFO]  -------------------- [  Processing aborted.  ] --------------------
2022-11-25 13:19:04 !! [ERROR]  Errors Map: key(gitlab-ce-docker/default-Create) -> value(3 error(s) decoding:

* 'httpPort' expected type 'uint', got unconvertible type 'string', value: '30080'
* 'httpsPort' expected type 'uint', got unconvertible type 'string', value: '30443'
* 'sshPort' expected type 'uint', got unconvertible type 'string', value: '30022')
2022-11-25 13:19:04 !! [ERROR]  Apply failed => some error(s) occurred during plugins apply process.
```

- issue 1: `'httpPort' expected type 'uint', got unconvertible type 'string', value: '30080'`
- issue2: it looks like gitlab-ce-docker installation failed, why is the installation process not interrupted and the tools jenkins and harbor continue to be installed later? 

### How to Reproduce?

N/A

### Anything else

N/A

### DevStream Version

latest

--- a/internal/pkg/configmanager/app.go
+++ b/internal/pkg/configmanager/app.go
@@ -3,6 +3,9 @@ package configmanager
 import (
 	"fmt"
 
+	"gopkg.in/yaml.v3"
+
+	"github.com/devstream-io/devstream/pkg/util/file"
 	"github.com/devstream-io/devstream/pkg/util/scm"
 )
 
@@ -12,7 +15,7 @@ const (
 
 type repoTemplate struct {
 	*scm.SCMInfo `yaml:",inline"`
-	Vars         map[string]any `yaml:"vars"`
+	Vars         RawOptions `yaml:"vars"`
 }
 
 type app struct {
@@ -22,6 +25,31 @@ type app struct {
 	RepoTemplate *repoTemplate `yaml:"repoTemplate" mapstructure:"repoTemplate"`
 	CIRawConfigs []pipelineRaw `yaml:"ci" mapstructure:"ci"`
 	CDRawConfigs []pipelineRaw `yaml:"cd" mapstructure:"cd"`
+}
+
+func getAppsFromConfigFileWithVarsRendered(fileBytes []byte, vars map[string]any) ([]*app, error) {
+	yamlPath := "$.apps[*]"
+	yamlStrArray, err := file.GetYamlNodeArrayByPath(fileBytes, yamlPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if yamlStrArray == nil {
+		return make([]*app, 0), nil
+	}
+
+	yamlWithVars, err := renderConfigWithVariables(yamlStrArray.StrOrigin, vars)
+	if err != nil {
+		return nil, err
+	}
+
+	var retTApps = make([]*app, 0)
+	err = yaml.Unmarshal(yamlWithVars, &retTApps)
+	if err != nil {
+		return nil, err
+	}
+
+	return retTApps, nil
 }
 
 // getAppPipelineTool generate ci/cd tools from app config
@@ -57,11 +85,14 @@ func (a *app) getRepoTemplateTool() (*Tool, error) {
 		if err != nil {
 			return nil, fmt.Errorf("configmanager[app] parse repoTemplate failed: %w", err)
 		}
+		if a.RepoTemplate.Vars == nil {
+			a.RepoTemplate.Vars = make(RawOptions)
+		}
 		return newTool(
 			repoScaffoldingPluginName, a.Name, RawOptions{
 				"destinationRepo": RawOptions(appRepo.Encode()),
 				"sourceRepo":      RawOptions(templateRepo.Encode()),
-				"vars":            RawOptions(a.RepoTemplate.Vars),
+				"vars":            a.RepoTemplate.Vars,
 			},
 		), nil
 	}

--- a/internal/pkg/configmanager/app_test.go
+++ b/internal/pkg/configmanager/app_test.go
@@ -1,0 +1,27 @@
+package configmanager
+
+import (
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("getToolsFromConfigFileWithVarsRendered", func() {
+	const appsConfig = `---
+apps:
+- name: app-1
+  cd:
+  - type: template
+    vars:
+      app: [[ appName ]]
+`
+	When("get apps from config file", func() {
+		It("should return config with vars", func() {
+			apps, err := getAppsFromConfigFileWithVarsRendered([]byte(appsConfig), map[string]any{"appName": interface{}("app-1")})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(apps).NotTo(BeNil())
+			Expect(len(apps)).To(Equal(1))
+			Expect(len(apps[0].CDRawConfigs)).To(Equal(1))
+			Expect(apps[0].CDRawConfigs[0].Vars["app"]).To(Equal(interface{}("app-1")))
+		})
+	})
+})

--- a/internal/pkg/configmanager/config.go
+++ b/internal/pkg/configmanager/config.go
@@ -10,23 +10,19 @@ import (
 
 // Config is a general config in DevStream.
 type Config struct {
-	Config              CoreConfig         `yaml:"config"`
-	Vars                map[string]any     `yaml:"vars"`
-	Tools               Tools              `yaml:"tools"`
-	Apps                []app              `yaml:"apps"`
-	PipelineTemplates   []pipelineTemplate `yaml:"pipelineTemplates"`
-	pipelineTemplateMap map[string]string  `yaml:"-"`
+	Config CoreConfig     `yaml:"config"`
+	Vars   map[string]any `yaml:"vars"`
+	Tools  Tools          `yaml:"tools"`
+	Apps   []*app         `yaml:"apps"`
+	// We'll read the pipeline templates from config file and render it to pipelineTemplateMap in no time.
+	//PipelineTemplates   []*pipelineTemplate `yaml:"-"`
+	pipelineTemplateMap map[string]string `yaml:"-"`
 }
 
 func (c *Config) getToolsFromApps() (Tools, error) {
-	err := c.renderPipelineTemplateMap()
-	if err != nil {
-		return nil, err
-	}
-
 	var tools Tools
 	for _, a := range c.Apps {
-		appTools, err := c.getToolsWithVarsFromApp(a)
+		appTools, err := c.getToolsFromApp(a)
 		if err != nil {
 			return nil, err
 		}
@@ -35,51 +31,33 @@ func (c *Config) getToolsFromApps() (Tools, error) {
 	return tools, nil
 }
 
-func (c *Config) getToolsWithVarsFromApp(a app) (Tools, error) {
-	appStr, err := yaml.Marshal(a)
+func (c *Config) getToolsFromApp(a *app) (Tools, error) {
+	// generate app repo and template repo from scmInfo
+	a.setDefault()
+	repoScaffoldingTool, err := a.getRepoTemplateTool()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("app[%s] get repo failed: %w", a.Name, err)
 	}
 
-	// 1. render appStr with Vars
-	appRenderStr, err := renderConfigWithVariables(string(appStr), c.Vars)
+	// get ci/cd pipelineTemplates
+	appVars := a.Spec.merge(c.Vars)
+	tools, err := a.generateCICDToolsFromAppConfig(c.pipelineTemplateMap, appVars)
 	if err != nil {
-		log.Debugf("configmanager/app %s render globalVars %+v failed", appRenderStr, c.Vars)
-		return nil, fmt.Errorf("app render globalVars failed: %w", err)
-	}
-
-	// 2. unmarshal app config for render pipelineTemplate
-	var rawApp app
-	if err = yaml.Unmarshal(appRenderStr, &rawApp); err != nil {
-		return nil, fmt.Errorf("app parse yaml failed: %w", err)
-	}
-
-	// 3. generate app repo and template repo from scmInfo
-	rawApp.setDefault()
-	repoScaffoldingTool, err := rawApp.getRepoTemplateTool()
-	if err != nil {
-		return nil, fmt.Errorf("app[%s] get repo failed: %w", rawApp.Name, err)
-	}
-
-	// 4. get ci/cd pipelineTemplates
-	appVars := rawApp.Spec.merge(c.Vars)
-	tools, err := rawApp.generateCICDToolsFromAppConfig(c.pipelineTemplateMap, appVars)
-	if err != nil {
-		return nil, fmt.Errorf("app[%s] get pipeline tools failed: %w", rawApp.Name, err)
+		return nil, fmt.Errorf("app[%s] get pipeline tools failed: %w", a.Name, err)
 	}
 	if repoScaffoldingTool != nil {
 		tools = append(tools, repoScaffoldingTool)
 	}
 
-	// 5. all tools from apps should depend on the original tools,
-	//    because dtm will execute all original tools first, then execute all tools from apps
+	// all tools from apps should depend on the original tools,
+	// because dtm will execute all original tools first, then execute all tools from apps
 	for _, toolFromApps := range tools {
 		for _, t := range c.Tools {
 			toolFromApps.DependsOn = append(toolFromApps.DependsOn, t.KeyWithNameAndInstanceID())
 		}
 	}
 
-	log.Debugf("Have got %d tools from app %s.", len(tools), rawApp.Name)
+	log.Debugf("Have got %d tools from app %s.", len(tools), a.Name)
 	for i, t := range tools {
 		log.Debugf("Tool %d: %v", i+1, t)
 	}
@@ -87,40 +65,11 @@ func (c *Config) getToolsWithVarsFromApp(a app) (Tools, error) {
 	return tools, nil
 }
 
-func (c *Config) renderToolsWithVars() error {
-	toolsStr, err := yaml.Marshal(c.Tools)
-	if err != nil {
-		return err
-	}
-
-	toolsStrWithVars, err := renderConfigWithVariables(string(toolsStr), c.Vars)
-	if err != nil {
-		return err
-	}
-
-	var tools Tools
-	if err = yaml.Unmarshal(toolsStrWithVars, &tools); err != nil {
-		return err
-	}
-	c.Tools = tools
-
-	return nil
-}
-
-func (c *Config) renderPipelineTemplateMap() error {
-	c.pipelineTemplateMap = make(map[string]string)
-	for _, pt := range c.PipelineTemplates {
-		tplBytes, err := yaml.Marshal(pt)
-		if err != nil {
-			return err
-		}
-		c.pipelineTemplateMap[pt.Name] = string(tplBytes)
-	}
-	return nil
-}
-
 func (c *Config) renderInstanceIDtoOptions() {
 	for _, t := range c.Tools {
+		if t.Options == nil {
+			t.Options = make(RawOptions)
+		}
 		t.Options["instanceID"] = t.InstanceID
 	}
 }

--- a/internal/pkg/configmanager/configmanager.go
+++ b/internal/pkg/configmanager/configmanager.go
@@ -1,12 +1,8 @@
 package configmanager
 
 import (
+	"fmt"
 	"os"
-	"regexp"
-
-	"gopkg.in/yaml.v3"
-
-	"github.com/devstream-io/devstream/pkg/util/log"
 )
 
 // Manager is used to load the config file from the ConfigFilePath and finally get the Config object.
@@ -28,7 +24,7 @@ func NewManager(configFilePath string) *Manager {
 // 3. Validation.
 func (m *Manager) LoadConfig() (*Config, error) {
 	// step 1
-	c, err := m.getConfigFromFile()
+	c, err := m.getConfigFromFileWithGlobalVars()
 	if err != nil {
 		return nil, err
 	}
@@ -39,10 +35,6 @@ func (m *Manager) LoadConfig() (*Config, error) {
 		return nil, err
 	}
 
-	err = c.renderToolsWithVars()
-	if err != nil {
-		return nil, err
-	}
 	c.Tools = append(c.Tools, appTools...)
 	c.renderInstanceIDtoOptions()
 
@@ -54,26 +46,54 @@ func (m *Manager) LoadConfig() (*Config, error) {
 	return c, nil
 }
 
-// getConfigFromFile gets Config from the config file specified by Manager.ConfigFilePath
-func (m *Manager) getConfigFromFile() (*Config, error) {
+// getConfigFromFileWithGlobalVars gets Config from the config file specified by Manager.ConfigFilePath, then:
+// 1. render the global variables to Config.Tools and Config.Apps
+// 2. transfer the PipelineTemplates to Config.pipelineTemplateMap, it's map[string]string type.
+// We can't render the original config file to Config.PipelineTemplates directly for the:
+//   1. variables rendered must be before the yaml.Unmarshal() called for the [[ foo ]] will be treated as a two-dimensional array by the yaml parser;
+//   2. the variables used([[ foo ]]) in the Config.PipelineTemplates can be defined in the Config.Apps or Config.Vars;
+//   3. pipeline templates are used in apps, so it would be more appropriate to refer to pipeline templates when dealing with apps
+func (m *Manager) getConfigFromFileWithGlobalVars() (*Config, error) {
 	configBytes, err := os.ReadFile(m.ConfigFilePath)
 	if err != nil {
 		return nil, err
 	}
 
-	configBytesEscaped := escapeBrackets(configBytes)
-
-	var c Config
-	if err = yaml.Unmarshal(configBytesEscaped, &c); err != nil {
-		log.Errorf("Please verify the format of your config. Error: %s.", err)
-		return nil, err
+	// global variables
+	vars, err := getVarsFromConfigFile(configBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get variables from config file. Error: %w", err)
 	}
 
-	return &c, nil
-}
+	// tools with global variables rendered
+	tools, err := getToolsFromConfigFileWithVarsRendered(configBytes, vars)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tools from config file. Error: %w", err)
+	}
 
-// escapeBrackets is used to escape []byte(": [[xxx]]xxx\n") to []byte(": \"[[xxx]]\"xxx\n")
-func escapeBrackets(param []byte) []byte {
-	re := regexp.MustCompile(`([^:]+:)(\s*)((\[\[[^\]]+\]\][^\s\[]*)+)[^\s#\n]*`)
-	return re.ReplaceAll(param, []byte("$1$2\"$3\""))
+	// apps with global variables rendered
+	apps, err := getAppsFromConfigFileWithVarsRendered(configBytes, vars)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get apps from config file. Error: %w", err)
+	}
+
+	// pipelineTemplateMap transfer from PipelineTemplates
+	pipelineTemplateMap, err := getPipelineTemplatesMapFromConfigFile(configBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pipelineTemplatesMap from config file. Error: %w", err)
+	}
+
+	// coreConfig without any changes
+	coreConfig, err := getCoreConfigFromConfigFile(configBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get coreConfig from config file. Error: %w", err)
+	}
+
+	return &Config{
+		Config:              *coreConfig,
+		Vars:                vars,
+		Tools:               tools,
+		Apps:                apps,
+		pipelineTemplateMap: pipelineTemplateMap,
+	}, nil
 }

--- a/internal/pkg/configmanager/configmanager_test.go
+++ b/internal/pkg/configmanager/configmanager_test.go
@@ -17,7 +17,8 @@ config:
 
 vars:
   foo1: bar1
-  foo2: bar2
+  foo2: 123
+  appName: service-a
   registryType: dockerhub
   argocdNamespace: argocd
 
@@ -56,15 +57,17 @@ apps:
       destination:
         namespace: devstream-io
     vars: # optional, use to render vars in template（valid only if the cd.type is template）
-      app: service-a
+      app: [[ appName ]]
 
 tools:
 - name: plugin1
   instanceID: default
+  dependsOn: []
   options:
     foo1: [[ foo1 ]]
 - name: plugin2
   instanceID: tluafed
+  dependsOn: []
   options:
     foo: bar
     foo2: [[ foo2 ]]
@@ -114,7 +117,7 @@ var _ = Describe("LoadConfig", func() {
 		Options: RawOptions{
 			"instanceID": "tluafed",
 			"foo":        "bar",
-			"foo2":       "bar2",
+			"foo2":       123,
 		},
 	}
 
@@ -232,7 +235,7 @@ var _ = Describe("LoadConfig", func() {
 			}))
 
 			// vars
-			Expect(len(cfg.Vars)).To(Equal(4))
+			Expect(len(cfg.Vars)).To(Equal(5))
 			Expect(cfg.Vars["foo1"]).To(Equal("bar1"))
 
 			// tools
@@ -257,7 +260,7 @@ var _ = Describe("LoadConfig", func() {
 	})
 })
 
-var _ = Describe("getConfigFromFile", func() {
+var _ = Describe("getConfigFromFileWithGlobalVars", func() {
 	BeforeEach(func() {
 		tmpWorkDir = GinkgoT().TempDir()
 		err := os.WriteFile(filepath.Join(tmpWorkDir, "config.yaml"), []byte(configFileStr), 0644)
@@ -267,7 +270,7 @@ var _ = Describe("getConfigFromFile", func() {
 	When("get config from file", func() {
 		It("should return a config", func() {
 			mgr := NewManager(filepath.Join(tmpWorkDir, "config.yaml"))
-			cfg, err := mgr.getConfigFromFile()
+			cfg, err := mgr.getConfigFromFileWithGlobalVars()
 			Expect(err).NotTo(HaveOccurred())
 			Expect(cfg.Config.State.Backend).To(Equal("local"))
 			Expect(cfg.Vars["foo1"]).To(Equal("bar1"))
@@ -275,27 +278,6 @@ var _ = Describe("getConfigFromFile", func() {
 			Expect(cfg.Apps[0].Name).To(Equal("service-a"))
 			Expect(len(cfg.Tools)).To(Equal(2))
 			Expect(cfg.Tools[1].Name).To(Equal("plugin2"))
-		})
-	})
-})
-
-var _ = Describe("escapeBrackets", func() {
-	When("escape brackets", func() {
-		It("should works right", func() {
-			testMap := map[string]string{
-				"foo: [[ foo ]]":           "foo: \"[[ foo ]]\"",
-				"foo: [[ foo ]] #comment":  "foo: \"[[ foo ]]\" #comment",
-				"foo: xx[[ foo ]]":         "foo: xx[[ foo ]]",
-				"foo: [[ foo ]]xx":         "foo: \"[[ foo ]]xx\"",
-				"foo: [[ foo ]]/[[ poo ]]": "foo: \"[[ foo ]]/[[ poo ]]\"",
-				`foo: [[ test ]]
-poo: [[ gg ]]`: "foo: \"[[ test ]]\"\npoo: \"[[ gg ]]\"",
-			}
-
-			for testStr, expectStr := range testMap {
-				retStr1 := escapeBrackets([]byte(testStr))
-				Expect(string(retStr1)).Should(Equal(expectStr))
-			}
 		})
 	})
 })

--- a/internal/pkg/configmanager/coreconfig.go
+++ b/internal/pkg/configmanager/coreconfig.go
@@ -1,5 +1,11 @@
 package configmanager
 
+import (
+	"gopkg.in/yaml.v3"
+
+	"github.com/devstream-io/devstream/pkg/util/file"
+)
+
 type CoreConfig struct {
 	State *State `yaml:"state"`
 }
@@ -22,4 +28,19 @@ type StateConfigOptions struct {
 	// for k8s backend
 	Namespace string `yaml:"namespace"`
 	ConfigMap string `yaml:"configmap"`
+}
+
+func getCoreConfigFromConfigFile(fileBytes []byte) (*CoreConfig, error) {
+	yamlPath := "$.config"
+	yamlStr, err := file.GetYamlNodeStrByPath(fileBytes, yamlPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var config *CoreConfig
+	err = yaml.Unmarshal([]byte(yamlStr), &config)
+	if err != nil {
+		return nil, err
+	}
+	return config, nil
 }

--- a/internal/pkg/configmanager/coreconfig_test.go
+++ b/internal/pkg/configmanager/coreconfig_test.go
@@ -1,0 +1,25 @@
+package configmanager
+
+import (
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("getCoreConfigFromConfigFile", func() {
+	const configFile = `---
+config:
+  state:
+    backend: local
+    options:
+      stateFile: devstream.state
+`
+	When("get core config from config file", func() {
+		It("should works fine", func() {
+			cc, err := getCoreConfigFromConfigFile([]byte(configFile))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(cc).NotTo(BeNil())
+			Expect(cc.State.Backend).To(Equal("local"))
+			Expect(cc.State.Options.StateFile).To(Equal("devstream.state"))
+		})
+	})
+})

--- a/internal/pkg/configmanager/pipelinetemplate.go
+++ b/internal/pkg/configmanager/pipelinetemplate.go
@@ -3,6 +3,8 @@ package configmanager
 import (
 	"fmt"
 
+	"github.com/devstream-io/devstream/pkg/util/file"
+
 	"github.com/imdario/mergo"
 	"gopkg.in/yaml.v3"
 
@@ -23,6 +25,29 @@ type (
 		Options RawOptions `yaml:"options"`
 	}
 )
+
+func getPipelineTemplatesMapFromConfigFile(fileBytes []byte) (map[string]string, error) {
+	yamlPath := "$.pipelineTemplates[*]"
+	yamlStrArray, err := file.GetYamlNodeArrayByPath(fileBytes, yamlPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if yamlStrArray == nil {
+		return make(map[string]string, 0), nil
+	}
+
+	var retMap = make(map[string]string)
+	for _, tplStr := range yamlStrArray.StrArray {
+		tplName, err := file.GetYamlNodeStrByPath([]byte(tplStr), "$.name")
+		if err != nil {
+			return nil, err
+		}
+		retMap[tplName] = tplStr
+	}
+
+	return retMap, nil
+}
 
 // getPipelineTemplate will generate pipleinTemplate from pipelineRaw
 func (p *pipelineRaw) getPipelineTemplate(templateMap map[string]string, globalVars map[string]any) (*pipelineTemplate, error) {

--- a/internal/pkg/configmanager/pipelinetemplate_test.go
+++ b/internal/pkg/configmanager/pipelinetemplate_test.go
@@ -85,7 +85,6 @@ var _ = Describe("pipelineRaw struct", func() {
 			It("should return err", func() {
 				_, err := r.getPipelineTemplate(templateMap, globalVars)
 				Expect(err).Error().Should(HaveOccurred())
-				Expect(err.Error()).Should(ContainSubstring("render pipelineTemplate failed"))
 			})
 		})
 		When("template is not valid yaml format", func() {
@@ -239,6 +238,29 @@ var _ = Describe("PipelineTemplate struct", func() {
 					},
 				}))
 			})
+		})
+	})
+})
+
+var _ = Describe("getPipelineTemplatesMapFromConfigFile", func() {
+	const toolsConfig = `pipelineTemplates:
+- name: tpl1
+  type: github-actions
+  options:
+    foo: bar
+`
+	const pipelineTemplate = `  name: tpl1
+  type: github-actions
+  options:
+    foo: bar`
+
+	When("get tools from config file", func() {
+		It("should return config with vars", func() {
+			pipelineTemplatesMap, err := getPipelineTemplatesMapFromConfigFile([]byte(toolsConfig))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(pipelineTemplatesMap).NotTo(BeNil())
+			Expect(len(pipelineTemplatesMap)).To(Equal(1))
+			Expect(pipelineTemplatesMap["tpl1"]).To(Equal(pipelineTemplate))
 		})
 	})
 })

--- a/internal/pkg/configmanager/tool.go
+++ b/internal/pkg/configmanager/tool.go
@@ -9,6 +9,7 @@ import (
 	"gopkg.in/yaml.v3"
 
 	"github.com/devstream-io/devstream/internal/pkg/version"
+	"github.com/devstream-io/devstream/pkg/util/file"
 	"github.com/devstream-io/devstream/pkg/util/validator"
 )
 
@@ -43,6 +44,31 @@ func newTool(name, instanceID string, options RawOptions) *Tool {
 		DependsOn:  []string{},
 		Options:    options,
 	}
+}
+
+func getToolsFromConfigFileWithVarsRendered(fileBytes []byte, vars map[string]any) (Tools, error) {
+	yamlPath := "$.tools[*]"
+	yamlStrArray, err := file.GetYamlNodeArrayByPath(fileBytes, yamlPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if yamlStrArray == nil {
+		return make(Tools, 0), nil
+	}
+
+	yamlWithVars, err := renderConfigWithVariables(yamlStrArray.StrOrigin, vars)
+	if err != nil {
+		return nil, err
+	}
+
+	var retTools = make(Tools, 0)
+	err = yaml.Unmarshal(yamlWithVars, &retTools)
+	if err != nil {
+		return nil, err
+	}
+
+	return retTools, nil
 }
 
 func (t *Tool) String() string {

--- a/internal/pkg/configmanager/tool_test.go
+++ b/internal/pkg/configmanager/tool_test.go
@@ -139,3 +139,24 @@ var _ = Describe("Tool struct", func() {
 		})
 	})
 })
+
+var _ = Describe("getToolsFromConfigFileWithVarsRendered", func() {
+	const toolsConfig = `---
+tools:
+- name: name1
+  instanceID: ins-1
+  dependsOn: []
+  options:
+    foo: [[ foo ]]
+`
+	When("get tools from config file", func() {
+		It("should return config with vars", func() {
+			tools, err := getToolsFromConfigFileWithVarsRendered([]byte(toolsConfig), map[string]any{"foo": interface{}("bar")})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(tools).NotTo(BeNil())
+			Expect(len(tools)).To(Equal(1))
+			Expect(len(tools[0].Options)).To(Equal(1))
+			Expect(tools[0].Options["foo"]).To(Equal(interface{}("bar")))
+		})
+	})
+})

--- a/internal/pkg/configmanager/variables.go
+++ b/internal/pkg/configmanager/variables.go
@@ -1,11 +1,32 @@
 package configmanager
 
 import (
+	"gopkg.in/yaml.v3"
+
+	"github.com/devstream-io/devstream/pkg/util/file"
 	"github.com/devstream-io/devstream/pkg/util/log"
 	"github.com/devstream-io/devstream/pkg/util/template"
 )
 
+func getVarsFromConfigFile(fileBytes []byte) (map[string]any, error) {
+	yamlPath := "$.vars"
+	yamlStr, err := file.GetYamlNodeStrByPath(fileBytes, yamlPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var retMap = make(map[string]any)
+	err = yaml.Unmarshal([]byte(yamlStr), retMap)
+	if err != nil {
+		return nil, err
+	}
+	return retMap, nil
+}
+
 func renderConfigWithVariables(fileContent string, variables map[string]interface{}) ([]byte, error) {
+	if len(variables) == 0 {
+		return []byte(fileContent), nil
+	}
 	logFunc(fileContent, variables)
 
 	str, err := template.New().

--- a/internal/pkg/configmanager/variables_test.go
+++ b/internal/pkg/configmanager/variables_test.go
@@ -24,3 +24,23 @@ var _ = Describe("renderConfigWithVariables", func() {
 		})
 	})
 })
+
+var _ = Describe("getVarsFromConfigFile", func() {
+	const configFile = `---
+foo: bar
+vars:
+  foo1: bar1
+  foo2: 123
+foo3: bar3
+`
+	When("get vars from config file", func() {
+		It("should works fine", func() {
+			varMap, err := getVarsFromConfigFile([]byte(configFile))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(varMap).NotTo(BeNil())
+			Expect(len(varMap)).To(Equal(2))
+			Expect(varMap["foo1"]).To(Equal(interface{}("bar1")))
+			Expect(varMap["foo2"]).To(Equal(interface{}(123)))
+		})
+	})
+})

--- a/pkg/util/file/yaml.go
+++ b/pkg/util/file/yaml.go
@@ -48,9 +48,9 @@ func GetYamlNodeArrayByPath(content []byte, path string) (*YamlSequenceNode, err
 	if !ok {
 		return nil, fmt.Errorf("yaml parse path[%s] is not valid sequenceNode", string(content))
 	}
-	var nodeArray []string
-	for _, seqNode := range seqNode.Values {
-		nodeArray = append(nodeArray, seqNode.String())
+	var nodeArray = make([]string, 0)
+	for _, sn := range seqNode.Values {
+		nodeArray = append(nodeArray, sn.String())
 	}
 	y := &YamlSequenceNode{
 		StrOrigin: node.String(),

--- a/pkg/util/template/processor.go
+++ b/pkg/util/template/processor.go
@@ -4,7 +4,7 @@ import "regexp"
 
 // this is because our variables' syntax is [[ varName ]]
 // while Go's template is [[ .varName ]]
-func AddDotForVariablesInConfig(s string) string {
+func addDotForVariablesInConfig(s string) string {
 	// regex := `\[\[\s*(.*)\s*\]\]`
 	// r := regexp.MustCompile(regex)
 	// return r.ReplaceAllString(s, "[[ .$1 ]]")
@@ -15,11 +15,9 @@ func AddDotForVariablesInConfig(s string) string {
 
 func AddDotForVariablesInConfigProcessor() Processor {
 	return func(bytes []byte) ([]byte, error) {
-		return []byte(AddDotForVariablesInConfig(string(bytes))), nil
+		return []byte(addDotForVariablesInConfig(string(bytes))), nil
 	}
 }
-
-// Quick Calls
 
 func (r *rendererWithGetter) AddDotForVariablesInConfigProcessor() *rendererWithGetter {
 	return r.AddProcessor(AddDotForVariablesInConfigProcessor())

--- a/pkg/util/template/processor_test.go
+++ b/pkg/util/template/processor_test.go
@@ -7,21 +7,21 @@ import (
 )
 
 func TestAddDotForVariablesInConfigNormal(t *testing.T) {
-	res := AddDotForVariablesInConfig("[[varNameA]]")
+	res := addDotForVariablesInConfig("[[varNameA]]")
 	assert.Equal(t, "[[ .varNameA]]", res, "Adding dot for variable names passed.")
 }
 
 func TestAddDotForVariablesInConfigWithSpaces(t *testing.T) {
-	res := AddDotForVariablesInConfig("[[  varNameA]]")
+	res := addDotForVariablesInConfig("[[  varNameA]]")
 	assert.Equal(t, "[[ .varNameA]]", res, "Adding dot for variable names passed.")
 }
 
 func TestAddDotForVariablesInConfigWithTrailingSpaces(t *testing.T) {
-	res := AddDotForVariablesInConfig("[[ varNameA  ]]")
+	res := addDotForVariablesInConfig("[[ varNameA  ]]")
 	assert.Equal(t, "[[ .varNameA  ]]", res, "Adding dot for variable names passed.")
 }
 
 func TestAddDotForVariablesInConfigMultipleVars(t *testing.T) {
-	res := AddDotForVariablesInConfig("[[ varNameA ]]/[[ varNameB ]]/[[ varNameC ]]")
+	res := addDotForVariablesInConfig("[[ varNameA ]]/[[ varNameB ]]/[[ varNameC ]]")
 	assert.Equal(t, "[[ .varNameA ]]/[[ .varNameB ]]/[[ .varNameC ]]", res, "Adding dot for variable names passed.")
 }

